### PR TITLE
[SYCL-MLIR] Define `sycl.host.set_captured` op.

### PR DESCRIPTION
**Pull request by jopperm**

### CODE REVIEW

**Comment by jopperm** [comment]
Failing E2E test `SYCL :: Plugin/level_zero_batch_event_status.cpp` seems unrelated.

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td [hunk 1]
```diff
@@ -123,6 +123,24 @@ def SYCLHostHandlerSetNDRange : SYCL_HostOp<"handler.set_nd_range"> {
...
+def SYCLHostSetCaptured : SYCL_HostOp<"set_captured"> {
+  let summary = "Marks a value as captured by a kernel function object.";
+  let description = [{
+    This operation represents that the given `value` was captured in the kernel
+    function object `lambda` at index `index`. The op is created during the 
+    progressive raising towards the `sycl.host.launch_kernel` op.
+
+    Example:
+    ```
+    sycl.host.set_captured $lambda[2] = %accessor : !llvm.ptr, !llvm.ptr
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td:135]:
```suggestion
    sycl.host.set_captured %lambda[2] = %accessor : !llvm.ptr, !llvm.ptr
```

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td [hunk 2]
```diff
@@ -123,6 +123,24 @@ def SYCLHostHandlerSetNDRange : SYCL_HostOp<"handler.set_nd_range"> {
...
+    progressive raising towards the `sycl.host.launch_kernel` op.
+
+    Example:
+    ```
+    sycl.host.set_captured %lambda[2] = %accessor : !llvm.ptr, !llvm.ptr
+    ```
+  }];
+  let arguments = (ins LLVM_AnyPointer:$lambda, I64Attr:$index, AnyType:$value);
+  let assemblyFormat = [{
+    $lambda `[` $index `]` `=` $value attr-dict `:` type(operands)
```
Comment (sommerlukas) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td:140]: Nit: Do we need the type of the `lambda` argument in the assembly format? I would expect it to always be `!llvm.ptr` with opaque pointers.

**Review by sommerlukas — APPROVED** [review]

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td (hunk 2 above)
Comment (jopperm) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td:140]: Right, no, will remove it.

Comment (etiotto) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td:140]: Question: do we need to know whether the `value` was captured by value or reference ?

Comment (sommerlukas) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td:140]: Per SYCL specification, only capture by value is possible for kernel function objects.

**Review by victor-eds — APPROVED** [review]

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td (hunk 2 above)
Comment (jopperm) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td]: Done. I didn't see a buildable version or alias of the `!llvm.ptr` type upstream, so I introduced this here for now.

**Review by whitneywhtsang — APPROVED** [review]

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td (hunk 2 above)
Comment (jopperm) [reply mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLHostOps.td:140]: Discussed with @sommerlukas that leaving out the type is not worth the haphazard "extension" of the LLVM dialect for our dialect.

**Pull request merged**